broadinstitute/single_cell_portal_core
# Log web vitals and client hardware metrics to Mixpanel (SCP-3299)

This logs metrics on page-load performance -- known as [web vitals](https://web.dev/vitals/) -- and client hardware to Mixpanel via Bard.

Good web vitals improve user experience and search engine ranking.  They help assess impact of simple-to-implement-and-hard-to-quantify optimizations, like changing compression from gzip to brotli.  They are measured for all page views.

Tracked web vitals:
* TTFB: time to first byte, the earliest page load event
* FCP: first contentful paint, when users gets first visual feedback
* LCP: largest contentful paint, a core measure of perceived load time
* FID: first input delay, when user can first interact (e.g. click)
* CLS: cumulative layout shift, measures visual stability

LCP, FID, and CLS are core web vitals.  TTFB and FCP are useful supplements.

Since users' machines and connection speeds impact performance, this also adds tracking for client hardware:
* Memory (GB RAM)
* CPU (# CPUs)
* Internet connection: 
  * Type: effective connection type, e.g. "4g"
  * RTT: round trip time, i.e. latency / ping (ms)
  * Downlink: download bandwidth in megabits per second (Mbps)

Client hardware is also tracked in any event that tracks `perfTime` -- e.g. fetching scatter plot data.  This could help, later perhaps, tailor how much data we send based on the client's connection speed.

Here's what it looks like in Mixpanel:
<img width="1014" alt="Web_vitals_and_client_hardware_properties_in_Mixpanel_SCP_2021-04-26" src="https://user-images.githubusercontent.com/1334561/116098673-42637500-a679-11eb-96cc-e4317dad5bfe.png">

To test:
* Open a page in SCP
* Go to another page
* (Note: cancelled web request is OK.  [`sendBeacon`](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/sendBeacon) would be ideal, but complicated for Bard to handle.)
* Go to [Live View in Mixpanel dev](https://mixpanel.com/report/2085496/view/19055/live)
* Search "web-vitals", verify that properties marked in screenshot appear

This satisfies SCP-3299.
